File: 2.20/scala/sol.scala
object SameParity {

  def isEven(x: Int): Boolean = {
    return x%2 == 1
  }

  def sameParity(xs: List[Int]): List[Int] = {
    if(xs.isEmpty)
      return Nil
    val x = xs.head
    
    if(isEven(x)){
      return xs.filter(isEven)
    }
    else{
      return xs.filter( (x: Int ) => !isEven(x) )
    }
  }
  def main(args: Array[String]): Unit = {
    println(sameParity(List(1, 2, 3, 4, 5, 6, 7)))
    println(sameParity(List(2, 3, 4, 5, 6, 7)))
  }
}

File: 2.20/haskell/sol.hs
isEven n = n `mod` 2 == 0

sameParity [] = []
sameParity (x:xs)
    | isEven x  = x : filter isEven xs
    | otherwise = x : filter (\x -> not $ isEven x) xs

main = do putStrLn $ show $ sameParity [1, 2, 3, 4, 5, 6, 7]
          putStrLn $ show $ sameParity [2, 3, 4, 5, 6, 7]
